File: top_hot100/0078Subsets/subsets.go
package subsets

func subsets(nums []int) [][]int {
	result := make([][]int, 0)
	emptySet := make([]int, 0)
	result = append(result, emptySet)
	getAll(nums, emptySet, &result)
	return result
}

func getAll(nums []int, oneSet []int, result *[][]int) {
	if len(nums) == 0 {
		return
	}
	for i, num := range nums {
		tempSet := make([]int, len(oneSet)+1)
		copy(tempSet, oneSet)
		tempSet[len(oneSet)] = num
		*result = append(*result, tempSet)
		getAll(nums[i+1:], tempSet, result)
	}
}

File: top_hot100/0078Subsets/subsets.py
# coding: utf-8

class Solution:
    def subsets(self, nums):
        result = [[]]
        self.getall(nums, [], result)
        return result
    
    def getall(self, nums, one_set, result):
        if len(nums) == 0:
            return 
        for index, num in enumerate(nums):
            temp = one_set[:] # copy一份新的数组
            temp.append(num)
            result.append(temp)
            self.getall(nums[index+1:], temp, result)